python/turbine_models/tests/vmfb_comparison.py

```diff
@@ -0,0 +1,183 @@
+
+    # Prepare the modules for the IREE runtime context
+    vm_modules = [mod, ireert.create_hal_module(config.vm_instance, config.device)]
+
+    # Include parameter module if external weight file is used
+    if external_weight_file:
+        param_module = ireert.create_io_parameters_module(
+            config.vm_instance, index.create_provider(scope="model")
+        )
+        vm_modules.insert(0, param_module)
+
+    # Create the system context with the given configuration and modules
+    ctx = ireert.SystemContext(vm_modules=vm_modules, config=config)
+
+    # Initialize the tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(
+        hf_model_name, use_fast=False, use_auth_token=hf_auth_token
+    )
+
+    # Convert the prompt to input tensor
+    initial_input = tokenizer(prompt, return_tensors="pt")
+    example_input_id = initial_input.input_ids
+    device_inputs = [ireert.asdevicearray(config.device, example_input_id)]
+
+    # Get the compiled module
+    ModuleCompiled = ctx.modules.state_update
+    results = ModuleCompiled["run_initialize"](*device_inputs)
+
+    def format_out(results):
+        # Convert the output to a PyTorch tensor
+        return torch.tensor(results.to_host()[0][0])
+
+    # Token generation loop
+    while True:
+        next_token_tensor = format_out(results)
+        yield next_token_tensor.item()  # Yield the scalar value of the tensor
+
+        # Run the next step of the model
+        results = ModuleCompiled["run_forward"](results)
+
+        # Check for the end-of-sentence token
+        if next_token_tensor.item() == tokenizer.eos_token_id and break_on_eos:
+            break
+
+
+def run_vmfb_comparison(
+    prompt,
+    hf_auth_token,
+    hf_model_name,
+    vmfb_path,
+    external_weight_file,
+    break_on_eos=True,
+):
+    # Initialize generators with the prompt and specific arguments
+    print("Using prompt:")
+    print(prompt)
+    torch_gen = torch_token_generator(
+        prompt=prompt,
+        hf_auth_token=hf_auth_token,
+        hf_model_name=hf_model_name,
+        break_on_eos=break_on_eos,
+    )
+
+    print(
+        "Generating Torch tokens... The pipeline needs to be initialized first so the first few tokens may take a while."
+    )
+    torch_tokens = list(tqdm(torch_gen, desc="Generating Torch tokens"))
+    del torch_gen
+
+    # Run turbine until an equal number of tokens has been generated
+    print(
+        "Generating Turbine tokens... The pipeline needs to be initialized first so the first few tokens may take a while."
+    )
+    turbine_gen = turbine_token_generator(
+        prompt=prompt,
+        hf_model_name=hf_model_name,
+        vmfb_path=vmfb_path,
+        external_weight_file=external_weight_file,
+        hf_auth_token=hf_auth_token,
+        break_on_eos=break_on_eos,
+    )
+    turbine_tokens = []
+    for _ in tqdm(range(len(torch_tokens)), desc="Generating Turbine tokens"):
```

Comment:
Why only provide the number of torch tokens?

Reply:
I've fixed the major things and applied Black. This should be ready to merge. I would like to have the test in the CI before address the rest of the problems, such that as I do the rest of the things I can keep track of when I broke stuff.

Thanks @IanNod for the comments. I intend to address them in a separate PR where i do all of the above + refactor things.

Todo items for tomorrow, in priority order.
Immediately:
- figure out iree target triple, and ireert.Config. @IanNod is there a GPU server I could ssh into to try it out?
Start the discussion about:
- generate more informative safename to include device, precision, quantization. Also, maybe store all the relevant files in a cache in somewhere like `~/.shark_turbine_cache` (similar to huggingface's cache)
- maybe talk about standardizing where we put the argparse stuff and make argparse match functions. Stella and Dan suggested that we always write the function first, and then use chatgpt to generate the argparse code. I think i prefer pasting this code towards the end of the file, close to the definition of the "main function" that's being exposed as an api
Double check:
- i think i already fixed the quantization problems in my latest commit which has the first passing test, but i'll figure that out.
- i think my commit also addresses the break_on_eos weirdness, but will double check tomorrow.